haystack/preview/components/samplers/top_p.py

```diff
@@ -0,0 +1,182 @@
+        :param score_field: The name of the field that should be used to store the scores in a document's metadata.
+        :param device: torch device (for example, cuda:0, cpu, mps) to limit inference to a specific device.
+        """
+        torch_and_transformers_import.check()
+        super().__init__()
+
+        self.model_name_or_path = model_name_or_path
+        self.top_p = top_p
+        self.score_field = score_field
+        self.device = device
+        self.model = None
+        self.tokenizer = None
+
+    def warm_up(self):
+        if self.model_name_or_path and (self.model is None and self.tokenizer is None):
+            self.model = AutoModelForSequenceClassification.from_pretrained(self.model_name_or_path)
+            self.model.eval()
+            self.tokenizer = AutoTokenizer.from_pretrained(self.model_name_or_path)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            top_p=self.top_p,
+            score_field=self.score_field,
+            device=self.device,
+            model_name_or_path=self.model_name_or_path,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "TopPSampler":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    @component.output_types(documents=List[Document])
+    def run(self, query: str, documents: List[Document], top_p: Optional[float] = None):
+        """
+        Returns a list of documents filtered using `top_p`, based on the similarity scores between the query and the
+        documents whose cumulative probability is less than or equal to `top_p`.
+
+        :param query: Query string.
+        :param documents: List of Documents.
+        :param top_p: Cumulative probability threshold for filtering the documents. If not provided, the top_p value
+        set during TopPSampler initialization is used.
+        :return: List of Documents sorted by (desc.) similarity with the query.
+        """
+        if top_p is None:
+            top_p = self.top_p if self.top_p else 1.0
+
+        if not documents:
+            return []
+
+        if self.model_name_or_path and self.model is None:
+            raise ComponentError(
+                f"The component {self.__class__.__name__} not warmed up. Run 'warm_up()' before calling 'run()'."
+            )
+        if not self.model_name_or_path and not self.has_scores(documents):
+            raise ComponentError(
+                f"The component {self.__class__.__name__} requires scores in the documents' metadata. "
+                f"Either set 'model_name_or_path' to scored documents or add scores to the documents."
+            )
+
+        needs_scoring = self.model_name_or_path and not self.has_scores(documents)
```

Comment:
I think `needs_scoring` could be False if the `model_name_or_path` is None and if `not self.has_scores` is True, which would lead to an error which might be hard for the user to understand.